controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -203,32 +203,27 @@ public void initializeStreamWriters(final ClientFactory clientFactory,
                 .thenCompose(retentionSet -> {
                     StreamCutRecord latestCut = retentionSet.stream()
                             .max(Comparator.comparingLong(StreamCutRecord::getRecordingTime)).orElse(null);
-                    return checkGenerateStreamCut(scope, stream, context, policy, latestCut, recordingTime)
-                            .thenCompose(x -> truncate(scope, stream, policy, context, retentionSet, recordingTime));
+                    return checkGenerateStreamCut(scope, stream, context, latestCut, recordingTime)
+                            .thenCompose(newRecord -> truncate(scope, stream, policy, context, retentionSet, newRecord, recordingTime));
                 });
 
     }
 
-    private CompletableFuture<Void> checkGenerateStreamCut(String scope, String stream, OperationContext context,
-                                                           RetentionPolicy policy, StreamCutRecord latestCut, long recordingTime) {
-        switch (policy.getRetentionType()) {
-            case TIME:
-                if (latestCut == null || recordingTime - latestCut.getRecordingTime() > Duration.ofMinutes(Config.MINIMUM_RETENTION_FREQUENCY_IN_MINUTES).toMillis()) {
-                    return generateStreamCut(scope, stream, context)
-                            .thenCompose(newRecord ->
-                                    streamMetadataStore.addStreamCutToRetentionSet(scope, stream, newRecord, context, executor));
-                } else {
-                    return  CompletableFuture.completedFuture(null);
-                }
-            case SIZE:
-            default:
-                throw new NotImplementedException("Size based retention");
+    private CompletableFuture<StreamCutRecord> checkGenerateStreamCut(String scope, String stream, OperationContext context,
+                                                           StreamCutRecord latestCut, long recordingTime) {
+        if (latestCut == null || recordingTime - latestCut.getRecordingTime() >
+                Duration.ofMinutes(Config.MINIMUM_RETENTION_FREQUENCY_IN_MINUTES).toMillis()) {
```

Comment:
Can you store this in a private final field so you don't need to compute this all the time?

Reply:
done